kube/src/api/params.rs

```diff
@@ -142,22 +142,44 @@ impl PostParams {
     }
 }
 
-/// Common query parameters for patch calls
-#[derive(Default, Clone)]
-pub struct PatchParams {
-    /// Whether to run this as a dry run
-    pub dry_run: bool,
-    /// Strategy which will be used. Defaults to [`PatchStrategy::Strategic`].
-    pub patch_strategy: PatchStrategy,
-    /// force Apply requests. Applicable only to [`PatchStrategy::Apply`].
-    pub force: bool,
-    /// fieldManager is a name of the actor that is making changes. Required for [`PatchStrategy::Apply`]
-    /// optional for everything else.
-    pub field_manager: Option<String>,
+/// Describes changes that should be applied to a resource
+///
+/// For all strategies except `Json`, patch can be represented with arbitrary
+/// serializable value, such as `serde_json::Value`. You may also want to use
+/// a `k8s-openapi` definition for the resource for the better type safety.
+#[non_exhaustive]
+pub enum Patch<T = ()> {
+    /// [Server side apply](https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply)
+    ///
+    /// Requires kubernetes >= 1.16
+    Apply {
+        /// fieldManager to use for Apply patches.
+        field_manager: String,
+        /// force Apply patches.
+        force: bool,
+        /// Patch itself
+        patch: T,
+    },
+    /// [JSON patch](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#use-a-json-merge-patch-to-update-a-deployment)
+    Json {
+        /// Patch itself
+        patch: json_patch::Patch,
```

Comment:
This forces people to import `json_patch` to write straight json patches which they may not want. It's good sanity checking on our end to have it, but thinking we perhaps should have an `TryInto<json_patch::Patch>` here instead if that works.

Reply:
> TryInto<json_patch::Patch> 

I think it does not have benefits over plain `serde_json::Value` because both things are fallible :(

I think we have two ways forward:

1. Take `json_patch::Patch`. Very comfortable for those who create patches using that library, but uncomfortable for users of other libraries.
2. Take `serde_json::Value`. Untyped, but works for everyone and does not pull excessive dependencies for those who do not use `json_patch` but construct patches manually.

I don't have a strong opinion on what strategy we should choose, I've never used json patches. Second looks more conservative.

Reply:
Is it a problem that it's fallible? I expected we would propagate the error if we have to convert it to a `Patch` and fail.

I guess if we feature flag jsonpatch then leaving it at 1 is probably a sound choice anyway as jsonpatch is already in the dependency tree.